File: solution/42. Trapping Rain Water/42_02.cpp
class Solution {
public:
    int trap(vector<int>& H) {
        int res = 0;
        stack<int> st;
        for (int i = 0; i < H.size(); ++i) {
            while (!st.empty() && H[st.top()] < H[i]) {
                int base = H[st.top()]; st.pop();
                if (st.empty()) break;
                int h = min(H[st.top()], H[i]) - base;
                int w = i - st.top() - 1;
                res += h * w;
            }
            st.push(i);
        }
        return res;
    }
};

File: solution/42. Trapping Rain Water/42_02.py
class Solution:
    def trap(self, H: List[int]) -> int:
        res = 0
        st = []
        for i, h, in enumerate(H):
            while st and H[st[-1]] < H[i]:
                base = H[st.pop()]
                if st:
                    h = min(H[st[-1]], H[i]) - base
                    w = i - st[-1] - 1
                    res += h * w
            st.append(i)
        return res
